text/0000-autoref-copy.md

```diff
@@ -0,0 +1,200 @@
+0 + &1;
+&0 + 1;
+// But these don't:
+&&0 + 1;
+&&0 + &1;
+&&0 + &&1;
+```
+
+These interactions confuse both new and experienced users without providing
+any significant value. It's clear what's intended by `map.get(1)` or
+`vec.iter().filter(|x| x > 1)`. When users encounter these errors in practice,
+the only reasonable thing they can do is add `&` and `*` as necessary to make
+their code compile.
+
+This RFC seeks to address one particular variant of this problem: passing
+owned `Copy` data where a reference was expected.
+
+Example:
+
+```rust
+use std::collections::HashMap;
+fn main() {
+    let mut map = HashMap::new();
+    map.insert(1, "hello!");
+
+    println!("{:?}", map.get(1)); // ERROR: expected `&{integer}`
+
+    // Instead, users have to write this:
+    println!("{:?}", map.get(&1));
+}
+```
+
+This is an unnecessary frustration. This RFC proposes to prevent this issue
+by auto-referencing `Copy` types.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When calling a function which expects a reference to a `Copy` type, you can
+pass an owned `Copy` type and a reference will be created automatically:
+
+```rust
+fn print_num(x: &usize) { println!("{:?}", x); }
+
+fn main() {
+  let x = 5;
+  print_num(x);
+}
+```
+
+This is particularly convenient when working with some generic types whose
+methods take references as arguments:
+
+```rust
+use std::collections::HashMap;
+fn main() {
+    let mut map = HashMap::new();
+    map.insert(1, "hello!");
+
+    // We can write this:
+    println!("{:?}", map.get(1));
+
+    // Instead of having to write this:
+    println!("{:?}", map.get(&1));
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+`Copy` types will autoreference: at
+[coercion sites](https://github.com/rust-lang/rfcs/blob/master/text/0401-coercions.md#coercions),
+when a reference is expected, but an owned
+`Copy` type is found, an `&` will be automatically inserted.
+
+If the `Copy` value is a temporary, its lifetime will be promoted to meet the
+required lifetime of the reference (where possible).
+This behavior is specified in detail in
+[RFC 66](https://github.com/rust-lang/rust/issues/15023) and @nikomatsakis's
+[amendment](https://github.com/nikomatsakis/rfcs/blob/rfc66-amendment/text/0066-better-temporary-lifetimes.md).
+
+When it is not possible to promote the `Copy` value's lifetime to the lifetime
+required by the reference, a customized error will be issued:
+
+```rust
+struct u8Ref<'a>(&'a u8);
+
+fn new_static(x: u8) -> u8Ref<'static> {
+    u8Ref(x) // ERROR: borrowed value does not live long enough
+    //    ^autoreference occurs here, but `x` does not live long enough
+}
+```
+
+If the lifetime of the reference would overlap with a mutable reference or a
+mutation of the referenced value, a custom error will be issued:
+
+```rust
+struct u8Ref<'a>(&'a u8);
+
+fn main() {
+    let mut x = 5;
+    let y = u8Ref(x);
```

Comment:
I am torn about whether to treat this auto-ref as borrowing the actual variable. I too find it somewhat surprising; I think I might prefer that we always create a temporary (with the standard lifetime). It would certainly suffice for the cases that are motivating the RFC; it seems like those cases where it would not suffice are also kind of confusing and unexpected to me (i.e., a reference that lives longer than the current statement).

It also seems more compatible with possibly extending this "auto-ref" to types beyond `Copy` types. At least the way we had previously talked about that, we wanted `foo(x)` to be *loosely* equivalent to `{ let y = x; foo(&y) }`, but with better error messages indicating that one can add an `&x` if one must retain ownership. I am still interested in thinking that over, but it seems incompatible with the approach of this RFC.